MudDataGrid: Add Localization Capabilites

Meduris opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->
resolves #6099, 
resolves #6963, 
resolves #6838, 
resolves #6871

Use Microsoft.Extensions.Localization to provide translations for DataGrid filter texts

The intention is to provide a new Service, which loads the default english translation and can be replaced by 3rd partys to add custom translations.

The localization is split into two services, an internal one which also provides the english translation, and a second public service which can be implemented and registered by users to provide their own translations.

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
added new unit tests and checked visually

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

## Docs:
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/0da4664e-0c51-4721-a258-6b17a8bc6877)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/5fd624cd-e93b-469e-9adc-cedfd4a2837e)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/6bc22ca6-a5ee-4b60-85be-3bb8693900b4)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/c4cba6c7-e2dc-45ae-be74-5e0a319c6f56)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/83819b76-ffbb-42f8-b884-6f5d93ec3cee)

ScarletKuro commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor:
```diff
@@ -4,6 +4,8 @@
 @using MudBlazor.Utilities
 @using MudBlazor.Extensions
 
+@inject DataGridLocalizer Localizer
```
Although this is a draft, I have a question: Why is the inclusion of the `DataGridLocalizer` layer necessary? Wouldn't it be more suitable to inject `IStringLocalizer` or `IStringLocalizer<DataGridLocalization>` instead?

Meduris replied:
The additional wrapper allows 3rd partys to override the ``DataGridLocalizer`` Service and supply their translations this way.
Using ``IStringLocalizer<DataGridLocalization>`` directly would not allow this, as it would be hard coded to the resx file provided with MudBlazor (at least as far as I am aware).

If the intention is to merge translations into this project, this is not a problem.

Meduris replied:
Though I'd gladly consider other approaches on how we can get 3rd party translation support

ScarletKuro replied:
> way. Using `IStringLocalizer<DataGridLocalization>` directly would not allow this, as it would be hard coded to the resx file provided with MudBlazor (at least as far as I am aware).

It would not? `IStringLocalizer<T>` just inherits `IStringLocalizer` and nothing else.
If you implement your own `IStringLocalizer` let's say `JsonStringLocalizer`.

<details>
  <summary>JsonStringLocalizer.cs example</summary>

  
  ```CSharp
public class JsonStringLocalizer : IStringLocalizer
{
    private readonly IDistributedCache _cache;
    private readonly JsonSerializer _serializer = new JsonSerializer();
    public JsonStringLocalizer(IDistributedCache cache)
    {
        _cache = cache;
    }
    public LocalizedString this[string name]
    {
        get
        {
            string value = GetString(name);
            return new LocalizedString(name, value ?? name, value == null);
        }
    }
    public LocalizedString this[string name, params object[] arguments]
    {
        get
        {
            var actualValue = this[name];
            return !actualValue.ResourceNotFound
                ? new LocalizedString(name, string.Format(actualValue.Value, arguments), false)
                : actualValue;
        }
    }
    public IEnumerable<LocalizedString> GetAllStrings(bool includeParentCultures)
    {
        string filePath = $"Resources/{Thread.CurrentThread.CurrentCulture.Name}.json";
        using (var str = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
        using (var sReader = new StreamReader(str))
        using (var reader = new JsonTextReader(sReader))
        {
            while (reader.Read())
            {
                if (reader.TokenType != JsonToken.PropertyName)
                    continue;
                string key = (string)reader.Value;
                reader.Read();
                string value = _serializer.Deserialize<string>(reader);
                yield return new LocalizedString(key, value, false);
            }
        }
    }
    private string GetString(string key)
    {
        string relativeFilePath = $"Resources/{Thread.CurrentThread.CurrentCulture.Name}.json";
        string fullFilePath = Path.GetFullPath(relativeFilePath);
        if (File.Exists(fullFilePath))
        {
            string cacheKey = $"locale_{Thread.CurrentThread.CurrentCulture.Name}_{key}";
            string cacheValue = _cache.GetString(cacheKey);
            if (!string.IsNullOrEmpty(cacheValue)) return cacheValue;
            string result = GetValueFromJSON(key, Path.GetFullPath(relativeFilePath));
            if (!string.IsNullOrEmpty(result)) _cache.SetString(cacheKey, result);
            return result;
        }
        return default(string);
    }
    private string GetValueFromJSON(string propertyName, string filePath)
    {
        if (propertyName == null) return default;
        if (filePath == null) return default;
        using (var str = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
        using (var sReader = new StreamReader(str))
        using (var reader = new JsonTextReader(sReader))
        {
            while (reader.Read())
            {
                if (reader.TokenType == JsonToken.PropertyName && (string)reader.Value == propertyName)
                {
                    reader.Read();
                    return _serializer.Deserialize<string>(reader);
                }
            }
            return default;
        }
    }
}
  ```
</details>

And now if you implement your own `IStringLocalizerFactory` that will create the `JsonStringLocalizer` and register it in DI:
```
services.AddLocalization();
services.AddDistributedMemoryCache();
services.AddSingleton<IStringLocalizerFactory, JsonStringLocalizerFactory>();
```
Then the `IStringLocalizer` will use your `JsonStringLocalizer` not the default `ResourceManagerStringLocalizer`.

ScarletKuro replied:
And the `IStringLocalizer<DataGridLocalization>` was a bad example from me, because the `T` should be the type to provide strings for, in this case `IStringLocalizer<MudDataGrid>`

Meduris replied:
That would work, but wouldn't this also break all other uses of IStringLocalizer that don't use that specific Localizer/Resource File?

To my understanding, your example would only return JsonStringLocalizer objects. So any component used would now just get JsonStringLocalizer instances, even if they'd use a resx file, dictionary or any other implementation normally.

Also referencing e.g. ``IStringLocalizer<DataGridLocalization>`` would still be hardwired to the ``DataGridLocalization`` class in the MudBlazor assembly.

Meduris replied:
> And the IStringLocalizer<DataGridLocalization> was a bad example from me, because the T should be the type to provide strings for, in this case IStringLocalizer<MudDataGrid>

I completely missed this comment at first thanks to github.

So we'd rather want a system, where you do not reference a specific resource file, but rather just specify the component, you want translations for?

Then we definitely need a custom StringLocalizer/Factory implementation as the default one uses the ``T`` to reference the resx file. We still kinda need a way to tell it which translation files to use.

I'll think about this and report back if I find an appropriate solution.

ScarletKuro replied:
Upon reviewing the implementation of `ResourceManagerStringLocalizer`, I realized that it requires the generic `IStringLocalizer` version, where `T` represents the type of the **resource class**, such as `DataGridLocalization`, rather than the type of the component. Initially, I assumed that providing any `T` would still work and the `ResourceManagerStringLocalizer` scans the entire assembly space and tries to locate the resource. The example here https://learn.microsoft.com/en-us/aspnet/core/fundamentals/localization/make-content-localizable?view=aspnetcore-7.0 confused me.

My original plan / idea was to directly utilize `IStringLocalizer` in the library. By default, we would utilize the `.resx` files and provide English translations only. If customers wish to add new translations, such as French, they create their own `DataGridLocalization.fr.resx` file(as example), populate it with the necessary translations, and it should work. In case they prefer to read translations from JSON or an online service(?), they just replace the default implementation with their own. However, this replacement would also require them to recreate the default(English) translations in their chosen format, i.e. this approach would be applied globally, meaning that both MudBlazor and the client application would need to read from the same format and that you can't do that MudBlazor reads from `.resx` and client application from `.json`.

Meduris replied:
Oh, I can totally see where the confusion came from... they used that as an example for something that doesn't provide translations where it automatically falls back to returning the translation key (which is why using the english translation as key might be the preffered solution)

I'll look into way I can build this so that the english translation does not need to be provided by the user if they just want to supply additional languages.
This most likely requires two services, an internal one for the translation itself and a public service which can be used by the user to provide the custom translations.
The idea for the internal service would be that it handles the english translation and uses the user provided translations for anything else. That way, it also wouldn't matter if we use 2 completely different ``IStringLocalizer`` implementations for english and the custom translations.

Should changing the default english text even be possible for external code?

just a quick pseudo implementation:
```c#
internal class InternalLocalizer
{
    UserProvidedMudBlazorLocalizer _localizer;

    public InternalLocalizer(UserProvidedMudBlazorLocalizer loc)
    {
        _localizer = loc;
    }

    public LocalizedString this[string key] => CurrentThread.CurrentUICulture == 'en' ? key : _localizer[key];
}

public class UserProvidedMudBlazorLocalizer
{
    public virtual LocalizedString this[string key] => key;
}
```

this way, we'd only directly use the ``InternalLocalizer``, possibly without even using a resx file inside of MudBlazor and users of MudBlazor can extend and register ``UserProvidedMudBlazorLocalizer`` and can choose whether they want to use resx, json or anything else for their translations.
If it should be possible to override english too, I'll have to test this with a few languages and check how it behaves if the default language provided by a user is not english

ScarletKuro replied:
> Should changing the default english text even be possible for external code?

I believe not. I think it's unnecessary complication, what do you think @henon?

Meduris replied:
Doing it like in my pseudo implementation would not allow for english texts to be changed.

The localizer would then be used like this ``@Localizer["Add Filter"]``  without having to add any form of translation files to MudBlazor. Though it would totally be possible to still use a resx or json file for the english translation.

I guess I'll wait for feedback on that idea and go ahead with it if you guys are ok with that solution.

ScarletKuro replied:
> The localizer would then be used like this `@Localizer["Add Filter"]` without having to add any form of translation files to MudBlazor. Though it would totally be possible to still use a resx or json file for the English translation.

I believe it would be valuable to seek a second opinion on this matter, @henon.
I get the idea and it's decent in general. However, I have some hesitations because I personally lean towards having a translation file (such as a resx or json file) for the sake of easier management and to avoid the need for searching through the source code later.
I think in future we would also want to provide the ability to localize the exception messages,  If my memory serves me correctly, there were some users who expressed a desire for this feature. It would be easier if it was in the file?

I don't have many specific requirements regarding the way the localization should function. Only two key aspects I'd like to emphasize: it should follow standard practices and be as simple as possible. I prefer to avoid reinventing the wheel, specifically referring to the `ITextLocalizer` used in Blazorise, which appears to be a derivative of the `IStringLocalizer` concept. However, creating a custom faced around the `IStringLocalizer` is acceptable, as it adheres to the standard approach. Thus, your initial draft seems suitable and aligns with my preferences.

Meduris replied:
> However, I have some hesitations because I personally lean towards having a translation file (such as a resx or json file) for the sake of easier management and to avoid the need for searching through the source code later.

Got it. That's not a problem with either solution.
Whether I can use the initial approach or the one I proposed in the comments, mostly depends on how we want to handle the english translations as my initial approach would require users to also provide english localization if they want to add languages.
Or the user would have to implement something like in ``InternalLocalizer`` to redirect english translations back to MudBlazor, so I'd prefer to implement the approach from here: https://github.com/MudBlazor/MudBlazor/pull/7024#discussion_r1231554465

Would you prefer to just use the english text as translation keys or a combination of the component name and text as I did use it in the pr?
(Basically ``"Add Filter"`` vs ``"MudDataGrid.AddFilter"``)

henon replied:
> > The localizer would then be used like this `@Localizer["Add Filter"]` without having to add any form of translation files to MudBlazor. Though it would totally be possible to still use a resx or json file for the English translation.
> 
> I believe it would be valuable to seek a second opinion on this matter, @henon. [...]

Yes, I also prefer all translations in one file, in my applications I use a Json file.

henon replied:
> > Should changing the default english text even be possible for external code?
> 
> I believe not. I think it's unnecessary complication, what do you think @henon?

Yes, I think it is not necessary, this is a use-case that probabyl will never happen.

ScarletKuro replied:
> Would you prefer to just use the english text as translation keys or a combination of the component name and text as I did use it in the pr? (Basically `"Add Filter"` vs `"MudDataGrid.AddFilter"`)

I like more the `MudDataGrid.AddFilte`.

henon replied:
I agree, it may duplicate a few translations but on the other hand there will be no interference between components.

haslingerm commented on src/MudBlazor/Resources/DataGridLocalization.resx:
I'm not the biggest fan of resx files, could you consider switching to json for example?
Also I'd prefer not one file per component, but one file per language, possibly with internal sections for each component (when creating a custom provider anyway that could improve structure).

So instead of `DataGridLocalization.resx` I would prefer `en.json` (shipped with Mudblazor) and then side-by-side other language files (e.g. `de.json`, `fr.json` etc.). In the file we could have a property "DataGrid" with an object value for the component's translations.
And then some logic to fall back to the original English translation if none is present for the locale. This would allow users to only override fields they need.
This would also allow for easier runtime changes based on (website) user's selection.

Maybe there will be a community sourced language pack in the future, but for the intial version it would be versatile at least.

ScarletKuro replied:
> I'm not the biggest fan of resx files, could you consider switching to json for example? Also I'd prefer not one file per component, but one file per language, possibly with internal sections for each component (when creating a custom provider anyway that could improve structure).
> 
> So instead of `DataGridLocalization.resx` I would prefer `en.json` (shipped with Mudblazor) and then side-by-side other language files (e.g. `de.json`, `fr.json` etc.). In the file we could have a property "DataGrid" with an object value for the component's translations. And then some logic to fall back to the original English translation if none is present for the locale. This would allow users to only override fields they need. This would also allow for easier runtime changes based on (website) user's selection.
> 
> Maybe there will be a community sourced language pack in the future, but for the intial version it would be versatile at least.

This comment highlights the importance of providing the flexibility to override the default implementation. Every user may have their own preferences and requirements, so it's ideal to give them the freedom to customize the behavior according to their needs. For example, if someone doesn't prefer using `.resx` files, they should have the ability to override the implementation and use `.json` files instead. Similarly, if they want to handle translations in a single file, they can create their own implementation that disregards the type and reads from the file based on `Thread.CurrentThread.CurrentCulture` / `Thread.CurrentThread.CurrentUICulture`.
I would totally put this on the users shoulders.

Regarding the default behavior, I believe it's beneficial to utilize `.resx` files since their support is already shipped with the `Microsoft.Extensions.Localization` library. By leveraging this existing infrastructure, we can minimize the amount of code and maintenance required on our part. It also provides a standard and widely recognized approach for localization in C#.

When it comes to organizing translations, both approaches have their advantages and disadvantages. Personally, I don't have a strong preference as it depends on the specific use case. However, I typically opt for the per-language approach.

haslingerm replied:
Leveraging the existing `.resx` infrastructure to reduce custom code in the library for this task I can understand even if I don't like it.

@Meduris I need this completed and in the framework for a project, so if you are interested in collab I'd be willing to pitch in.

Meduris replied:
@markushaslinger thanks for the offer, I am trying to finish this up today though. Just got held up by some other things and needed some time to cool off.

The solution I proposed in the other comments would just use ``.resx`` internally for the english translation and allow users of MudBlazor to implement their custom translations based on whatever they prefer. Only part I'd enforce for this is to still use the ``LocalizedString`` class which is part of ``Microsoft.Extensions.Localization`` as it allows to easily pass information about wheter a translation was found or not and I can use that to decide whether to fall back to the internal english translation or not.

I'll push an update soon which should include the first part of this, then most of the stuff left to do for me is to add the missing translations and fix the tests.

haslingerm replied:
@Meduris Awesome 👍 
Offer stands, if you need a second pair of hands or eyes shoot me a mention - I'll spend the day doing some refactoring anyway.

Meduris replied:
@markushaslinger I hope this solution works for you.

The ``InternalMudBlaozorLocalizer`` uses a resx file and the default ``IStringLocalizer`` for the english translation, but you can implement ``MudBlazorLocalizer`` however you want as long as you return a ``LocalizedString`` and it won't matter whatever way you use to provide your translations, be it resx or json or even just a plain hardcoded dictionary

ScarletKuro commented:
Side note: don't forget to add later the services to:
`src/MudBlazor.UnitTests/Generated/ApiDocsTests.cs`
`src/MudBlazor.UnitTests/Generated/ExampleDocsTests.cs`
so that the tests wouldn't fail.

Meduris commented:
> Side note: don't forget to add later the services to: `src/MudBlazor.UnitTests/Generated/ApiDocsTests.cs` `src/MudBlazor.UnitTests/Generated/ExampleDocsTests.cs` so that the tests wouldn't fail.

Thank you. That saves me a lot of time trying to find why the localizer was missing in the test setup.

Meduris commented:
Not sure what I should to about the coverage here, considering it complains about the localizer calls.

also with the current setup, users would have to register their custom localizer like this: ``services.AddTransient<MudBlazorLocalizer, CustomLocalizer>();``.

I tested this with a quick and dirty german translation and set the current ui cutlure to german and it successfully displayed the german texts. Deleting some of the keys in the german translation successfully used the default english text.

I guess, the two Localizer Services should have code documentation (even if the internal one is not used by users).

Also should the ``MudBlazorLocalizer`` be mentioned in a separate page for the web documentation or only on the MudDataGrid page for now?

Meduris commented:
One part I am not quite sure about so far is the localization of the filter operators, as I used a simple ``$"MudDataGrid.{operator}"`` for the key, which means keys for operators like ``is equals`` will contain the spaces.

I've been thinking about dynamically converting them to strings like ``IsEquals`` but I am not too sure about the overhead of something like this.

ScarletKuro commented:
> One part I am not quite sure about so far is the localization of the filter operators, as I used a simple `$"MudDataGrid.{operator}"` for the key, which means keys for operators like `is equals` will contain the spaces.
> 
> I've been thinking about dynamically converting them to strings like `IsEquals` but I am not too sure about the overhead of something like this.

Can't we just change this part
https://github.com/MudBlazor/MudBlazor/blob/dev/src/MudBlazor/Components/DataGrid/FilterOperator.cs
for example change this:
`public const string NotEmpty = "is not empty"`
to
`public const string NotEmpty = "is_not_empty"`
or whatever the naming style should be/
I don't think this is considered a breaking change? @henon

Meduris commented:
> Can't we just change this part?
https://github.com/MudBlazor/MudBlazor/blob/dev/src/MudBlazor/Components/DataGrid/FilterOperator.cs
for example change this:
public const string NotEmpty = "is not empty"
to
public const string NotEmpty = "is_not_empty"

that would also work if it doesn't break things. Then I'd just need a decision on which naming style to use. So far, I've been using the ``IsNotEmpty``, but I can switch to whatever style you prefer to use here

henon commented:
It might be breaking if somebody saved this string to save user preferences but I guess it is not too much of a problem? We can do it. The format (camel case or snake case) is not important I think.

henon commented:
If spaces aren't a problem I would just leave the values as they are now, that is the safest bet with regard to possible breaking changes. Nevermind how it looks in the resx file.
As for coverage, honestly I don't really see code that needs testing when I look a the changes. @ScarletKuro correct me if I am wrong.
 
Yes, we'd need a separate documentation page that explains how to localize a MudBlazor component. A very simple example would be enough for the start. We can expand the documentation later with more examples.

Meduris commented:
I noticed one special case with my current implementation at least for users using microsofts ``IStringLocalizer`` implementation:

If a user uses resx files the normal way and wants to have it use english as default fallback language, the user would have to leave the default culture variant of their resx files empty with how microsofts implementation works.
Otherwise, for localized strings present in the default culture resx file, it will always use those entries if the culture is not set to english.

Note, this is only about the fall back in place if e.g. CurrentCulture is set to ``de-de`` and ``userlang.de-de.resx`` doesn't contain a localization for a string but ``userlang.resx`` does. In this case, it would use the values from ``userlang.resx`` and only fall back to our included english translation, if neither ``userlang.de-de.resx`` nor ``userlang.resx`` contain a suitable localization.

I currently tend to leave it this way and document it properly, especially as this also makes it easier in case a user wants to provide a custom default language with resx based implementations. Also with custom implementations that don't use microsofts resx based code, this could also easily be done.


I hope its somewhat comprehensible what I am trying to get across here, it's been a long day for me.

ScarletKuro commented:
> As for coverage, honestly I don't really see code that needs testing when I look a the changes. @ScarletKuro correct me if I am wrong.

Yeah, it's mainly complaining about the razor file changes with `@Localizer["XYZ"]`
However, it would be a good idea to write small test for `InternalMudBlazorLocalizer` and `MudBlazorLocalizer`.

> the user would have to leave the default culture variant of their resx files empty with how microsofts implementation works

Wasn't this always the case for the `.resx`?

> I currently tend to leave it this way

Based on my understanding, I don't perceive any issues here. It seems that the only requirement is to ensure proper documentation, as you mentioned.

ScarletKuro commented:
Can you also add these two issues to "resolves": https://github.com/MudBlazor/MudBlazor/issues/6963, https://github.com/MudBlazor/MudBlazor/issues/6838

Meduris commented:
> Can you also add these three issues to "resolves": #6963, #6838, #6871

Done, though it seems for github to automatically close the issues after merging this, I'd need to write "resolves" before each single issue number. Let me know if I should do that

henon commented:
I linked the issues. So what is left to do for this PR?

henon commented:
Oh right, the doc sample

Meduris commented:
I just added some documentation

There are some things I still want to do though:
1. Include an auto generated list with all translation keys, so that it's easier to add custom translations
2. Add a short example for registering the custom localizer
3. If possible I'd prefer not to use razor components for the examples here, as it doesn't really make sense

Meduris committed:
MudDataGrid: Add Localization Capabilities

Meduris commented:
I got my last few points done and changed ``NullLoggerFactory`` to ``ILoggerFactory`` (the tests still use the ``NullLoggerFactory`` as those log a lot better what is actually going wrong.

Rebased on the current dev branch and squashed my commits

Meduris commented:
One thing I noticed where I am not quite sure what caused this is, that 2 times when starting the project, it still tried to auto generate the ``.html`` files in ``ManualMarkdown``. I don't see anything obvious I might have missed though.

ScarletKuro committed:
fix conflict with ServiceCollectionExtensions

ScarletKuro commented:
> One thing I noticed where I am not quite sure what caused this is, that 2 times when starting the project, it still tried to auto generate the `.html` files in `ManualMarkdown`. I don't see anything obvious I might have missed though.

Also hit this problem, this is concerning since 100% later someone going to PR with this files changed.
Are you sure the files should have `.html` extension? Looking at the `RTLLanguages`->`ManualMarkDown` it has .razor files.
Also why when I type in SearchBar the SubTitle doesn't show-up(`Localization allows to provide custom translations other then english.`)
Maybe @henon has a suggestion?

ScarletKuro commented:
Rest LGTM, I really like it. Just need to solve the problems above and it's good to go.

ScarletKuro committed:
Add readonly

henon commented:
Fixing the docs ...

henon committed:
Fix examples

henon committed:
Some typos, changes and improvements in the docs

henon commented:
@Meduris please pull. You can write the rest of the examples now. Basically, the files in the `Examples` dir ending in `...Eample.razor` are automatically compiled into the html files by our docs generator in the `Code` dir but the html files are not checked in.

henon commented:
> Also why when I type in SearchBar the SubTitle doesn't show-up(`Localization allows to provide custom translations other then english.`) Maybe @henon has a suggestion?

This is a shortcoming of our docs system. Subtitles in the search box are only supported for components with API pages ATM.

henon commented:
@Meduris I realize you put the code samples in the html files. They were supposed to be in the razor files. Fixing this also, because I deleted the htmls.

ScarletKuro commented on src/MudBlazor.Docs/Pages/Features/Localization/LocalizationPage.razor:
```diff
@@ -5,22 +5,30 @@
             <SectionHeader Title="Explanation">
                 <Description>
-                    MudBlazor itself provides english translation for texts found in e.g. the <CodeInline>MudDataGrid</CodeInline> filter options.
+                    MudBlazor itself provides English lanugage strings for texts found in e.g. the <CodeInline>MudDataGrid</CodeInline> filter options.
```
Typo lanugage instead of language

henon committed:
Revert "Fix examples"

This reverts commit 323d9940baa9dca4b30daba80fd0647312f80e6e.

henon committed:
Rename all manual markdown examples to NOT contain "Example"

henon commented:
> @Meduris I realize you put the code samples in the html files. They were supposed to be in the razor files. Fixing this also, because I deleted the htmls.

I am dumb. Didn't know how the manual markup system worked. You had it right except that the files should not have contained the term `Example` ;). Documentation looks awesome by the way.

Meduris commented:
Thank you for the input and the fixes.

> I am dumb. Didn't know how the manual markup system worked. You had it right except that the files should not have contained the term Example ;). Documentation looks awesome by the way.

I got this from Getting Started -> Installation, but if the single file solution like in RTLLanguages also works, that would be a solution too.

henon commented:
I don't think that that file in RTLLanugages is in use. I think it is a leftover because I can't find any use of that razor component in the docs.

Meduris commented:
Yeah, I can't find any uses either, and I think it would also kinda mess with the "copy code" button as either way, that is just going to copy the contents of the razor file as far as I understand it

henon commented:
Lol, you just found another bug in the doc system. When copying the snippet via the button I get this in the clipboard: `Snippet for component 'LocalizationServiceRegistration' not found!`

Meduris commented:
On a sidenote, the inline filters also have an issue when scrolling with a fixed header and will hide the header row after scrolling down

henon commented:
Okay, both are their own issues though, so I think this is ready for merge. You agree @Meduris ?

ScarletKuro committed:
GeneratedCodeAttribute ver 17

Meduris commented:
> Okay, both are their own issues though, so I think this is ready for merge. You agree @Meduris ?

From what I see, yes. I se nothing wrong with your edits and have nothing further to add for now

ScarletKuro commented:
I set generation version to 17, as its what VS 2022 will always set when opening solution, usually should not conflict with others IDE.

henon commented:
We'll merge and release tomorrow. Big thank you, outstanding work!

ScarletKuro commented:
> I don't think that that file in RTLLanugages is in use. I think it is a leftover because I can't find any use of that razor component in the docs.

Would be nice to cleanup all this.

Meduris commented:
> I set generation version to 17, as its what VS 2022 will always set when opening solution, usually should not conflict with others IDE.

yeah, doesn't seem to cause any issue with Rider at least. It doesn't even change the version back after adding a resource key.



> We'll merge and release tomorrow. Big thank you, outstanding work!

Thank you for all the positive feedback and your patience during the last week .

ScarletKuro commented:
@henon tbh for v7 breaking change I probably would still consider this change: 

> Can't we just change this part? https://github.com/MudBlazor/MudBlazor/blob/dev/src/MudBlazor/Components/DataGrid/FilterOperator.cs for example change this: `public const string NotEmpty = "is not empty"` to `public const string NotEmpty = "is_not_empty"` or whatever the naming style should be. I don't think this is considered a breaking change?

Simply for consistency of the strings + VS will complain with `The resource name 'MudDataGrid.=' is not valid identifier` even thought it still works, it's just not considered as standard.

Meduris commented:
> Simply for consistency of the strings + VS will complain with The resource name 'MudDataGrid.=' is not valid identifier even thought it still works, it's just not considered as standard.

we could also go the other way and change all keys to match this format: ``MudSataGrid.[english translation]``, though with the prefix it feels weird to have spaces in the keys.

Also noticed that for the operators like ``< > =`` it doesn't generate static references in the ``LanguageResource.Designer`` file. Should we change those texts to something it can generate that code for?

It doesn't really matter with how we use it though. The Localizer can still access them properly and we are not using those static references either way.

ScarletKuro commented:
> > Simply for consistency of the strings + VS will complain with `The resource name 'MudDataGrid.=' is not valid identifier` even thought it still works, it's just not considered as standard.

I don't believe it's necessary to modify the remaining translation keys solely because of the filter keys. In my opinion, the change in FilterOperator is a easy and straightforward and could be part of a major version update (because we should consider that users can be quite creative, and there might be cases where the operator values are utilized for unconventional purposes).

> It doesn't really matter with how we use it though. The Localizer can still access them properly and we are not using those static references either way.

Yeah, for now it's not a problem.

henon commented:
I agree about the plan to change the filter values in v7, adding it to our v7 project. Then making the release.

henon commented:
@Meduris we are finally releasing v7. Do you want to make the breaking change discussed at the end of this thread? If yes, please PR against dev.

ScarletKuro commented:
> @Meduris we are finally releasing v7. Do you want to make the breaking change discussed at the end of this thread? If yes, please PR against dev.

I think he is busy, since I had to finish https://github.com/MudBlazor/MudBlazor/pull/7389
